concrete-csprng/src/generators/aes_ctr/mod.rs

```diff
@@ -0,0 +1,223 @@
+//! A module implementing the random generator api with batched aes calls.
+//!
+//! This module provides a generic [`AesCtrGenerator`] structure which implements the
+//! [`super::RandomGenerator`] api using the AES block cipher in counter mode. That is, the
+//! generator holds a state which is incremented iteratively, to produce the stream of random
+//! values:
+//! ```ascii
+//!         state         state+1        state+2
+//!        ╔══↧══╗        ╔══↧══╗        ╔══↧══╗
+//!    key ↦ AES ║    key ↦ AES ║    key ↦ AES ║ ...
+//!        ╚══↧══╝        ╚══↧══╝        ╚══↧══╝
+//!        output0        output1        output2
+//!
+//!          t=0            t=1            t=2
+//! ```
+//!
+//! The [`AesCtrGenerator`] structure is generic over the AES block cipher, represented by the
+//! [`AesBlockCipher`] trait. Implementers only need to rewrite the `AesBlockCipher` bits, and can
+//! instantiate a full generator from the `AesCtrGenerator`.
+//!
+//! In the following section, we give details on the implementation of this generic generator.
+//!
+//! Coarse-grained pseudo-random table lookup
+//! =========================================
+//!
+//! To generate random values, we use the AES block cipher in counter mode. If we denote f the aes
+//! encryption function, we have:
+//! ```ascii
+//!     f: ⟦0;2¹²⁸ -1⟧ X ⟦0;2¹²⁸ -1⟧ ↦ ⟦0;2¹²⁸ -1⟧
+//!     f(secret_key, input) ↦ output
+//! ```
+//!
+//! If we fix the secret key to a value k, we have a pure function fₖ from ⟦0;2¹²⁸ -1⟧ to
+//! ⟦0;2¹²⁸-1⟧, transforming the state of the counter into a pseudo random value. Essentially, this
+//! fₖ function can be considered as a lookup function into a table of 2¹²⁸ pseudo-random values:
+//! ```ascii  
+//!     ╭──────────────┬──────────────┬─────┬──────────────╮
+//!     │     fₖ(0)    │     fₖ(1)    │     │  fₖ(2¹²⁸ -1) │
+//!     ╔═══════↧══════╦═══════↧══════╦═════╦═══════↧══════╗
+//!     ║┏━━━━━━━━━━━━┓║┏━━━━━━━━━━━━┓║     ║┏━━━━━━━━━━━━┓║
+//!     ║┃    u128    ┃║┃    u128    ┃║ ... ║┃    u128    ┃║
+//!     ║┗━━━━━━━━━━━━┛║┗━━━━━━━━━━━━┛║     ║┗━━━━━━━━━━━━┛║
+//!     ╚══════════════╩══════════════╩═════╩══════════════╝
+//! ```
+//!
+//! We call this input to the fₖ function, an _aes index_ of the pseudo-random table. The
```

Comment:
I added a note explaining that this is also called state or counter in the standard definitions, but I don't have ideas for simpler namings :confused:  . I used this one because it composes well with the explanation and the fact that there is also a byte index which together with the aes index form the table index, which _is the actual state_ of the generator (which yields bytes instead of integers of the size of aes_index, like in the standards). I would prefer not to change the `index` naming, as it would involve a long pass over the PR to ensure all sentences make sense.